### PR TITLE
Preferred Hotspots

### PR DESCRIPTION
**Pull request by ambuznego**
Reference: https://github.com/helium/router/issues/456

image
image
image
image
image

_________________________________________________

image
image
image
image

________________________________________________

```
➜  console git:(preferredHotspots) mix test

10:49:56.386 [info]  Migrations already up
......................................................................

Finished in 2.3 seconds (0.6s async, 1.7s sync)
70 tests, 0 failures
```

### CODE REVIEW

**Comment by vicmgs** [comment]
preferred status needs to be added to org import/export of organization_hotspots

File: lib/console_web/controllers/device_controller.ex [hunk 1]
```diff
@@ -58,6 +58,8 @@ defmodule ConsoleWeb.DeviceController do
         |> render("show.json", device: device)
       {:error, "Device limit reached"} ->
         {:error, :forbidden, "The device/organization cap has been met. To add devices or organizations for commercial use cases, reach out to sales@nova.xyz."}
+      {:error, %Ecto.Changeset{} = _} ->
+        {:error, :unprocessable_entity, "Invalid credentials"}
```
Comment (ambuznego) [inline lib/console_web/controllers/device_controller.ex:62]: fixing test error

File: assets/js/actions/apollo.js [hunk 2]
```diff
@@ -1,173 +1,179 @@
...
-              merge: false
+              merge: false,
             },
             allDevices: {
-              merge: false
+              merge: false,
             },
-            allMultiBuys: {
-              merge: false
+            allPacketConfigs: {
```
Comment (ambuznego) [inline assets/js/actions/apollo.js:123]: main change, rest is linting

File: lib/console/packet_configs/packet_config.ex [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    belongs_to :organization, Organization
+    has_many :devices, Device
+    has_many :labels, Label
+    timestamps()
+  end
+
+  def changeset(packet_config, attrs) do
+    attrs = Helpers.sanitize_attrs(attrs, ["name"])
+    IO.inspect attrs
```
Comment (vicmgs) [inline lib/console/packet_configs/packet_config.ex:27]: log

File: lib/console_web/controllers/router/device_controller.ex [hunk 4]
```diff
@@ -79,33 +80,52 @@ defmodule ConsoleWeb.Router.DeviceController do
...
+                  %{
+                    multi_buy_value: 1,
+                    preferred_hotspots: []
+                  }
                 _ ->
-                  max_value = Enum.max(label_multi_buys)
-                  case device.multi_buy_id do
-                    nil -> if max_value == 21, do: 9999, else: max_value
-                    _ -> if device.multi_buy.value == 21, do: 9999, else: device.multi_buy.value
+                  device_label = Labels.get_latest_applied_device_label(device.id, Enum.map(labels_with_packet_configs, fn l -> l.id end))
```
Comment (vicmgs) [inline lib/console_web/controllers/router/device_controller.ex:94]: this means the latest label will apply, but if the latest label doesnt have any packet settings, other labels with packet settings will be ignored, which isnt what we want probably.

**Comment by vicmgs** [comment]
overall works great!

File: lib/console_web/controllers/router/device_controller.ex (hunk 4 above)
Comment (ambuznego) [reply lib/console_web/controllers/router/device_controller.ex:94]: I am looking within `labels_with_config_profiles` which all have a config profile. from those, i'm pulling the latest.

**Comment by vicmgs** [comment]
import/export of orgs will also need to be accounted for before merge

**Comment by Jhobean** [comment]
Any ETA on this Commit? I'll love to use it asap. 😁 It seem a really good work.

**Pull request merged**